sophos/client.go

```diff
@@ -0,0 +1,276 @@
+	var err error
+	a := &SophosAdapter{
+		conf:   conf,
+		ctx:    context.Background(),
+		doStop: utils.NewEvent(),
+	}
+
+	a.uspClient, err = uspclient.NewClient(conf.ClientOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	a.httpClient = &http.Client{
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 10 * time.Second,
+			}).Dial,
+		},
+	}
+
+	a.chStopped = make(chan struct{})
+
+	a.wgSenders.Add(1)
+	go a.fetchEvents(eventsURL)
+
+	go func() {
+		a.wgSenders.Wait()
+		close(a.chStopped)
+	}()
+
+	return a, a.chStopped, nil
+}
+
+func (a *SophosAdapter) Close() error {
+	a.conf.ClientOptions.DebugLog("closing")
+	a.doStop.Set()
+	a.wgSenders.Wait()
+	err1 := a.uspClient.Drain(1 * time.Minute)
+	_, err2 := a.uspClient.Close()
+	a.httpClient.CloseIdleConnections()
+
+	if err1 != nil {
+		return err1
+	}
+
+	return err2
+}
+
+func (a *SophosAdapter) fetchEvents(url string) {
+	defer a.wgSenders.Done()
+	defer a.conf.ClientOptions.DebugLog(fmt.Sprintf("fetching of %s events exiting", url))
+
+	lastCursor := ""
+	has_more := "false"
+	for !a.doStop.WaitFor(30 * time.Second) {
+		// The makeOneRequest function handles error
+		// handling and fatal error handling.
+		items, newCursor, has_more_resp := a.makeOneRequest(url, lastCursor, has_more)
+		lastCursor = newCursor
+		has_more = has_more_resp
+		if items == nil {
+			continue
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err == nil {
+					continue
+				}
+				a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+				a.doStop.Set()
+				return
+			}
+		}
+	}
+}
+
+func (a *SophosAdapter) getJwt() string {
+
+	auth_url := "https://id.sophos.com/api/v2/oauth2/token"
+	method := "POST"
+
+	payload := strings.NewReader(fmt.Sprintf("grant_type=client_credentials&scope=token&client_id=%s&client_secret=%s", a.conf.ClientId, a.conf.ClientSecret))
+
+	client := &http.Client{}
+	req, err := http.NewRequest(method, auth_url, payload)
+
+	if err != nil {
+		fmt.Println(err)
+		return ""
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		fmt.Println(err)
+		return ""
+	}
+	defer resp.Body.Close()
+
+	// Evaluate if success.
+	if resp.StatusCode != http.StatusOK {
+		body, _ := ioutil.ReadAll(resp.Body)
+		//	a.conf.ClientOptions.DebugLog(fmt.Sprintf("response: %s", body))
+		//	a.conf.ClientOptions.DebugLog(fmt.Sprintf("error code: %s", resp.StatusCode))
+
+		a.conf.ClientOptions.OnWarning(fmt.Sprintf("sophos api non-200: %s\nRESPONSE: %s", resp.Status, string(body)))
```

Comment:
```suggestion
		a.conf.ClientOptions.OnError(fmt.Sprintf("sophos api non-200: %s\nRESPONSE: %s", resp.Status, string(body)))
```
Should this be an Error since if it happens the rest of the Adapter won't work and should just stop?